File: agent_sdk/agent.py
"""Agent v1 — claims auditor on the Claude Agent SDK.

An autonomous tool loop: the model is given the four audit tools as an in-process
MCP server and drives itself, calling ``fetch_claim`` / ``lookup_rule`` /
``check_patient_history`` to gather evidence and ``emit_finding`` (validated
against the ``Finding`` JSON Schema) to report each defect. Orchestration is the
model's job here; contrast ``agent_graph`` where it is an explicit graph.

Requires the optional ``sdk`` extra (``pip install -e ".[sdk]"``), the Claude
Code CLI on PATH, and ``ANTHROPIC_API_KEY``. The class raises a clear error if
the SDK is missing so the light-weight core/CI path never imports it.
"""

from __future__ import annotations

import asyncio
import json
import sys

from claims_audit.models import Claim, Finding
from claims_audit.prompting import SYSTEM_PROMPT, audit_task_prompt
from claims_audit.rules import RuleSet, load_rules
from claims_audit.tools import ToolContext, check_patient_history, fetch_claim, lookup_rule
from evals.harness import Usage

DEFAULT_MODEL = "claude-sonnet-5"
DEFAULT_MAX_TURNS = 12
DEFAULT_THINKING_TOKENS = 4000


def _text_result(payload: dict) -> dict:
    """Wrap a JSON payload in the SDK tool-result content shape."""
    return {"content": [{"type": "text", "text": json.dumps(payload)}]}


class ClaudeSDKAuditAgent:
    """AuditAgent implemented on claude-agent-sdk."""

    def __init__(
        self,
        claims: list[Claim],
        ruleset: RuleSet | None = None,
        model: str = DEFAULT_MODEL,
        thinking: bool = False,
        max_turns: int = DEFAULT_MAX_TURNS,
        provider: str | None = None,
    ):
        self.ruleset = ruleset or load_rules()
        self.ctx = ToolContext.build(claims, self.ruleset)
        # Resolve provider so the CLI can be pointed at OpenRouter's
        # Anthropic-compatible endpoint (experimental) via env passthrough.
        from providers.chat import build_config

        self._cfg = build_config(model=model, thinking=thinking, provider=provider)
        self.provider = self._cfg.provider
        self.model = self._cfg.model
        self.thinking = thinking
        self.max_turns = max_turns
        self.name = "claude-agent-sdk"
        self._usage = Usage()

    # ---- harness interface ------------------------------------------------

    def audit(self, claim: Claim) -> list[Finding]:
        # A single failed claim yields no findings rather than crashing a long
        # multi-claim run (the CLI spawns a subprocess per claim and can flake).
        try:
            return asyncio.run(self._audit_async(claim))
        except Exception as exc:  # noqa: BLE001 - resilience over strictness here
            print(f"[claude-agent-sdk] claim {claim.claim_id} errored: {exc}", file=sys.stderr)
            return []

    def usage(self) -> Usage:
        return Usage(self._usage.input_tokens, self._usage.output_tokens)

    # ---- SDK plumbing -----------------------------------------------------

    def _build_server(self):
        from claude_agent_sdk import create_sdk_mcp_server, tool

        ctx = self.ctx

        @tool("fetch_claim", "Fetch a full claim record by id.", {"claim_id": str})
        async def _fetch(args):
            return _text_result(fetch_claim(ctx, args["claim_id"]))

        @tool(
            "lookup_rule",
            "Look up rules by rule_id, by CPT code, or all rules if neither is given.",
            {"rule_id": str, "code": str},
        )
        async def _lookup(args):
            return _text_result(
                lookup_rule(ctx, rule_id=args.get("rule_id"), code=args.get("code"))
            )

        @tool(
            "check_patient_history",
            "Return the patient's other claims for cross-claim checks.",
            {"patient_id": str, "exclude_claim_id": str},
        )
        async def _history(args):
            return _text_result(
                check_patient_history(
                    ctx, args["patient_id"], args.get("exclude_claim_id")
                )
            )

        @tool(
            "emit_finding",
            "Emit one validated audit finding citing rule_id and line_refs.",
            {
                "claim_id": str,
                "rule_id": str,
                "defect_type": str,
                "line_refs": list,
                "severity": str,
                "rationale": str,
            },
        )
        async def _emit(args):
            payload = {
                "claim_id": args.get("claim_id"),
                "rule_id": args.get("rule_id"),
                "defect_type": args.get("defect_type"),
                "line_refs": args.get("line_refs") or [],
                "severity": args.get("severity") or "medium",
                "rationale": args.get("rationale") or "",
            }
            try:
                model = Finding.model_validate(payload)
            except Exception as exc:  # noqa: BLE001 - surface validation to model
                return {
                    "content": [{"type": "text", "text": f"REJECTED: {exc}"}],
                    "is_error": True,
                }
            ctx.emitted.append(model)
            return _text_result({"accepted": True})

        return create_sdk_mcp_server(
            name="audit", tools=[_fetch, _lookup, _history, _emit]
        )

    def _options(self, server):
        from claude_agent_sdk import ClaudeAgentOptions

        kwargs = dict(
            mcp_servers={"audit": server},
            allowed_tools=[
                "mcp__audit__fetch_claim",
                "mcp__audit__lookup_rule",
                "mcp__audit__check_patient_history",
                "mcp__audit__emit_finding",
            ],
            system_prompt=SYSTEM_PROMPT,
            model=self.model,
            max_turns=self.max_turns,
            # tools must be used offline: no filesystem / network side effects.
            setting_sources=None,
        )
        if self.thinking:
            kwargs["max_thinking_tokens"] = DEFAULT_THINKING_TOKENS

        # OpenRouter routing: point the Claude Code CLI at OpenRouter's
        # Anthropic-compatible endpoint. The CLI appends "/v1/messages" to
        # ANTHROPIC_BASE_URL, so the base must NOT already end in /v1 — OpenRouter
        # serves it at https://openrouter.ai/api/v1/messages. Auth is a bearer
        # token (ANTHROPIC_AUTH_TOKEN), which is what OpenRouter expects.
        if self.provider == "openrouter" and self._cfg.api_key:
            base = (self._cfg.base_url or "https://openrouter.ai/api/v1").rstrip("/")
            if base.endswith("/v1"):
                base = base[:-3].rstrip("/")
            kwargs["env"] = {
                "ANTHROPIC_BASE_URL": base,
                "ANTHROPIC_AUTH_TOKEN": self._cfg.api_key,
            }
        return ClaudeAgentOptions(**kwargs)

    async def _audit_async(self, claim: Claim) -> list[Finding]:
        from claude_agent_sdk import ResultMessage, query

        self.ctx.reset_emitted()
        server = self._build_server()
        options = self._options(server)
        prompt = audit_task_prompt(claim, self.ruleset)

        async for message in query(prompt=prompt, options=options):
            if isinstance(message, ResultMessage) and message.usage:
                u = message.usage
                self._usage = Usage(
                    self._usage.input_tokens + int(u.get("input_tokens", 0)),
                    self._usage.output_tokens + int(u.get("output_tokens", 0)),
                )
        # Only findings about THIS claim (guard against stray tool calls).
        return [f for f in self.ctx.emitted if f.claim_id == claim.claim_id]

File: agent_aisdk/src/agent.ts
/**
 * Agent v3 — the same auditor on the Vercel AI SDK.
 *
 * Like `agent_sdk` (and unlike the `agent_graph` fixed graph), the model drives
 * an autonomous tool loop: it is handed the four tools and calls `fetch_claim` /
 * `lookup_rule` / `check_patient_history` to gather evidence and `emit_finding`
 * per defect. Control flow lives in the model; the loop is bounded by
 * `stopWhen: stepCountIs(maxSteps)`. This deliberately mirrors the autonomous
 * `claude-agent-sdk` arm rather than the LangGraph arm, so the three-way
 * comparison contrasts genuinely different orchestration styles.
 */
import { type LanguageModel, type ToolSet, generateText, stepCountIs } from 'ai';
import type { AuditContext } from './context.js';
import type { Finding } from './models.js';
import { auditTaskPrompt, SYSTEM_PROMPT } from './prompting.js';
import { type ProviderConfig, providerOptions } from './provider.js';
import { buildTools } from './tools.js';

export const DEFAULT_MAX_STEPS = 12;

export interface ClaimUsage {
  input_tokens: number;
  output_tokens: number;
}

export interface ClaimAuditResult {
  claim_id: string;
  findings: Finding[];
  usage: ClaimUsage;
  latency_ms: number;
  steps: number;
  error?: string;
}

export interface AuditOptions {
  cfg: ProviderConfig;
  maxSteps?: number;
  /** Per-claim wall-clock ceiling; a wedged call becomes "no findings". */
  timeoutMs?: number | undefined;
  /**
   * Extra tools (e.g. discovered from an external MCP server) exposed to the
   * model alongside the four audit tools. Not used by the scored benchmark —
   * present so an MCP-sourced tool can be genuinely reachable by the agent.
   */
  extraTools?: ToolSet | undefined;
}

/**
 * Audit one claim. Never throws: a model/transport failure (or a per-claim
 * timeout) is recorded as zero findings with an `error` note, so one bad claim
 * cannot crash a multi-claim run — the same resilience the `claude-agent-sdk`
 * arm applies with its per-claim timeout.
 */
export async function auditClaim(
  ctx: AuditContext,
  claimId: string,
  model: LanguageModel,
  opts: AuditOptions,
): Promise<ClaimAuditResult> {
  ctx.resetEmitted();
  ctx.currentClaimId = claimId;

  const claim = ctx.claims.get(claimId);
  const start = performance.now();
  if (claim === undefined) {
    return {
      claim_id: claimId,
      findings: [],
      usage: { input_tokens: 0, output_tokens: 0 },
      latency_ms: performance.now() - start,
      steps: 0,
      error: `claim ${claimId} not found`,
    };
  }

  const po = providerOptions(opts.cfg);
  try {
    const result = await generateText({
      model,
      system: SYSTEM_PROMPT,
      prompt: auditTaskPrompt(claim, ctx.ruleset),
      tools: { ...buildTools(ctx), ...(opts.extraTools ?? {}) },
      maxOutputTokens: opts.cfg.maxTokens,
      stopWhen: stepCountIs(opts.maxSteps ?? DEFAULT_MAX_STEPS),
      ...(po ? { providerOptions: po } : {}),
      ...(opts.timeoutMs ? { abortSignal: AbortSignal.timeout(opts.timeoutMs) } : {}),
    });
    return {
      claim_id: claimId,
      // Only findings about THIS claim (guard against a stray emit).
      findings: ctx.emitted.filter((f) => f.claim_id === claimId),
      usage: {
        input_tokens: result.usage.inputTokens ?? 0,
        output_tokens: result.usage.outputTokens ?? 0,
      },
      latency_ms: performance.now() - start,
      steps: result.steps.length,
    };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      claim_id: claimId,
      findings: ctx.emitted.filter((f) => f.claim_id === claimId),
      usage: { input_tokens: 0, output_tokens: 0 },
      latency_ms: performance.now() - start,
      steps: 0,
      error: message,
    };
  }
}
